src/hosting/config.ts

```diff
@@ -0,0 +1,245 @@
+  if (!onlyString) {
+    return configs;
+  }
+
+  let onlyTargets = onlyString.split(",");
+  // If an unqualified "hosting" is in the --only,
+  // all hosting sites should be deployed.
+  if (onlyTargets.includes("hosting")) {
+    return configs;
+  }
+
+  // Strip out Hosting deploy targets from onlyTarget
+  onlyTargets = onlyTargets
+    .filter((target) => target.startsWith("hosting:"))
+    .map((target) => target.replace("hosting:", ""));
+
+  return matchingConfigs(configs, onlyTargets, /* assertMatch= */ true);
+}
+
+/**
+ * Returns a subset of configs that match the except string;
+ */
+export function filterExcept(configs: HostingMultiple, exceptOption?: string): HostingMultiple {
+  if (!exceptOption) {
+    return configs;
+  }
+
+  const exceptTargets = exceptOption.split(",");
+  if (exceptTargets.includes("hosting")) {
+    return [];
+  }
+
+  const exceptValues = exceptTargets
+    .filter((t) => t.startsWith("hosting:"))
+    .map((t) => t.replace("hosting:", ""));
+  const toReject = matchingConfigs(configs, exceptValues, /* assertMatch= */ false);
+
+  return configs.filter((c) => !toReject.find((r) => c.site === r.site && c.target === r.target));
+}
+
+/**
+ * Verifies that input in firebase.json is sane
+ * @param options options from the command library
+ * @return a deep copy of validated configs
+ */
+export function extract(options: HostingOptions): HostingMultiple {
+  const config = options.config.src;
+  if (!config.hosting) {
+    return [];
+  }
+  const assertOneTarget = (config: HostingSingle): void => {
+    if (config.target && config.site) {
+      throw new FirebaseError(
+        `Hosting configs should only include either "site" or "target", not both.`
+      );
+    }
+  };
+
+  if (!Array.isArray(config.hosting)) {
+    // Upgrade the type because we pinky swear to ensure site exists as a backup.
+    const res = cloneDeep(config.hosting) as unknown as RequireAtLeastOne<{
+      site: string;
+      target: string;
+    }>;
+    // earlier the default RTDB instance was used as the hosting site
+    // because it used to be created along with the Firebase project.
+    // RTDB instance creation is now deferred and decoupled from project creation.
+    // the fallback hosting site is now filled in through requireHostingSite.
+    if (!res.target && !res.site) {
+      // Fun fact. Site can be the empty string if someone just downloads code
+      // and launches the emulator before configuring a project.
+      res.site = options.site;
+    }
+    assertOneTarget(res);
+    return [res];
+  } else {
+    config.hosting.forEach(assertOneTarget);
+    return cloneDeep(config.hosting);
+  }
+}
+
+/** Validates hosting configs for semantic correctness. */
+export function validate(configs: HostingMultiple, options: HostingOptions): void {
+  for (const config of configs) {
+    validateOne(config, options);
+  }
+}
+
+function validateOne(config: HostingMultiple[number], options: HostingOptions): void {
```

Comment:
What does `[number]` do in this context? I've never seen this before (and you can imagine google searching "typescript `[number]`" doesn't exactly give me the right results haha)

Reply:
Type arithmetic. What happens when you index a HostingMultiple with a number? Since HostingMultiple is an array type, you get the element type of HostingMultiple. You would _think_ that would be a HostingSingle, but HostingSingle allows both site and target to be undefined and the array syntax of HostingMultiple always requires site or target to be defined.

Reply:
Oh, fun fact, you can also index by a literal string. For example `{ foo: string }["foo"]` is `string`. I think I do that elsewhere to say a `HostingOptions["duration"]` without separately exporting the duration field's type.